snapcraft/internal/project_loader/_parts_config.py

```diff
@@ -169,12 +172,25 @@ def load_part(self, part_name, plugin_name, part_properties):
                                                plugin_name,
                                                part_properties))
 
+        sources = getattr(plugin, 'PLUGIN_STAGE_SOURCES', None)
+        stage_packages_repo = repo.Repo(
+            plugin.packagedir, sources=sources,
+            project_options=self._project_options)
+
+        grammar_processor = grammar_processing.PartGrammarProcessor(
```

Comment:
ah I see now, you just execute the logic once in the plugin. There are benefits to this I guess as the repo is not setup during project load, I guess this is why you went down this path now.

Reply:
This should be a pattern that everything using grammar follows though.